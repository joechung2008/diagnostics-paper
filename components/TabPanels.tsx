import React, { useCallback, useRef } from "react";
import { Animated, ScrollView, StyleSheet, Text, View } from "react-native";
import { Gesture, GestureDetector } from "react-native-gesture-handler";
import { ActivityIndicator, useTheme } from "react-native-paper";
import BuildInfo from "./BuildInfo";
import Extension from "./Extension";
import Extensions from "./Extensions";
import ServerInfo from "./ServerInfo";
import type { Diagnostics, ExtensionInfo, KeyedNavLink } from "../lib/types";
import {
  createFlingDismissAnimation,
  createFlingResetAnimation,
  createFlingStartAnimation,
} from "../lib/animations";

interface TabPanelsProps {
  diagnostics: Diagnostics | undefined;
  error?: Error;
  extension: ExtensionInfo | undefined;
  isLoading: boolean;
  selectedTab: string;
  viewMode: "list" | "details";
  onLinkClick: (item: KeyedNavLink) => void;
  onViewModeChange: (mode: "list" | "details") => void;
}

const TabPanels: React.FC<TabPanelsProps> = ({
  diagnostics,
  error,
  extension,
  isLoading,
  selectedTab,
  viewMode,
  onLinkClick,
  onViewModeChange,
}) => {
  const theme = useTheme();

  // Animation values for fling feedback
  const { current: flingOpacity } = useRef(new Animated.Value(1));
  const { current: flingScale } = useRef(new Animated.Value(1));
  const { current: flingTranslateX } = useRef(new Animated.Value(0));
  const animationStarted = useRef(false);

  const resetFlingAnimation = useCallback(() => {
    createFlingResetAnimation(
      flingOpacity,
      flingScale,
      flingTranslateX
    ).start();
    // https://www.epicreact.dev/why-you-shouldnt-put-refs-in-a-dependency-array
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // Create swipe gestures using the modern Gesture API
  const flingGesture = Gesture.Pan()
    .onStart(() => {
      animationStarted.current = false;
      createFlingStartAnimation(flingOpacity, flingScale).start();
    })
    .onUpdate((event) => {
      if (!animationStarted.current) {
        let translateValue = 0;
        if (event.velocityX > 500 && event.translationX > 50) {
          translateValue = -200;
        } else if (event.velocityX < -500 && event.translationX < -50) {
          translateValue = 200;
        }
        if (translateValue !== 0) {
          animationStarted.current = true;
          createFlingDismissAnimation(flingTranslateX, translateValue).start();
        }
      }
    })
    .onEnd(() => {
      if (animationStarted.current) {
        onViewModeChange("list");
      }
    })
    .onFinalize(() => {
      resetFlingAnimation();
    });

  return (
    <ScrollView
      contentContainerStyle={isLoading || error ? styles.center : undefined}
      style={styles.content}
    >
      {isLoading ? (
        <ActivityIndicator
          accessibilityLabel="Loading diagnostics..."
          size="large"
        />
      ) : error ? (
        <Text style={[styles.errorText, { color: theme.colors.error }]}>
          Error loading diagnostics: {error.message}
        </Text>
      ) : selectedTab === "extensions" && diagnostics?.extensions ? (
        <View style={styles.tabContent}>
          <Animated.View
            style={[
              styles.gestureContainer,
              {
                opacity: flingOpacity,
                transform: [
                  { scale: flingScale },
                  { translateX: flingTranslateX },
                ],
              },
            ]}
          >
            {viewMode === "list" ? (
              <Extensions
                extensions={diagnostics.extensions}
                onLinkClick={onLinkClick}
              />
            ) : (
              <GestureDetector gesture={flingGesture}>
                {extension && <Extension {...extension} />}
              </GestureDetector>
            )}
          </Animated.View>
        </View>
      ) : selectedTab === "build" && diagnostics?.buildInfo ? (
        <View style={styles.tabContent}>
          <BuildInfo {...diagnostics.buildInfo} />
        </View>
      ) : selectedTab === "server" && diagnostics?.serverInfo ? (
        <View style={styles.tabContent}>
          <ServerInfo {...diagnostics.serverInfo} />
        </View>
      ) : null}
    </ScrollView>
  );
};

const styles = StyleSheet.create({
  center: {
    alignItems: "center",
    flex: 1,
    justifyContent: "center",
  },
  content: {
    flex: 1,
  },
  errorText: {
    textAlign: "center",
  },
  gestureContainer: {
    flex: 1,
  },
  tabContent: {
    padding: 10,
  },
});

export default TabPanels;
