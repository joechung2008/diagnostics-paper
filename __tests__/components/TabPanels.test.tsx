// Mock react-native-gesture-handler
import React from "react";
import { PaperProvider, MD3LightTheme } from "react-native-paper";
import { render } from "@testing-library/react-native";
import TabPanels from "../../components/TabPanels";

jest.mock("react-native-gesture-handler", () => ({
  Gesture: {
    Pan: jest.fn(() => ({
      onStart: jest.fn().mockReturnThis(),
      onUpdate: jest.fn().mockReturnThis(),
      onEnd: jest.fn().mockReturnThis(),
      onFinalize: jest.fn().mockReturnThis(),
    })),
  },
  GestureDetector: ({ children }: { children: React.ReactNode }) => children,
}));

describe("TabPanels", () => {
  const mockOnLinkClick = jest.fn();
  const mockOnViewModeChange = jest.fn();

  const mockDiagnostics = {
    buildInfo: { buildVersion: "1.0.0" },
    extensions: {
      ext1: {
        extensionName: "Extension One",
        manageSdpEnabled: true,
      },
    },
    serverInfo: {
      deploymentId: "test-deployment",
      extensionSync: { totalSyncAllCount: 5 },
      hostname: "test-host",
      serverId: "test-server",
    },
  };

  it("renders correctly in loading state", () => {
    const { toJSON } = render(
      <PaperProvider theme={MD3LightTheme}>
        <TabPanels
          diagnostics={undefined}
          extension={undefined}
          isLoading={true}
          selectedTab="extensions"
          viewMode="list"
          onLinkClick={mockOnLinkClick}
          onViewModeChange={mockOnViewModeChange}
        />
      </PaperProvider>
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it("renders correctly with error", () => {
    const mockError = new Error("Test error");

    const { toJSON } = render(
      <PaperProvider theme={MD3LightTheme}>
        <TabPanels
          diagnostics={undefined}
          error={mockError}
          extension={undefined}
          isLoading={false}
          selectedTab="extensions"
          viewMode="list"
          onLinkClick={mockOnLinkClick}
          onViewModeChange={mockOnViewModeChange}
        />
      </PaperProvider>
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it("renders correctly with extensions tab in list mode", () => {
    const { toJSON } = render(
      <PaperProvider theme={MD3LightTheme}>
        <TabPanels
          diagnostics={mockDiagnostics}
          extension={undefined}
          isLoading={false}
          selectedTab="extensions"
          viewMode="list"
          onLinkClick={mockOnLinkClick}
          onViewModeChange={mockOnViewModeChange}
        />
      </PaperProvider>
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it("renders correctly with build tab", () => {
    const { toJSON } = render(
      <PaperProvider theme={MD3LightTheme}>
        <TabPanels
          diagnostics={mockDiagnostics}
          extension={undefined}
          isLoading={false}
          selectedTab="build"
          viewMode="list"
          onLinkClick={mockOnLinkClick}
          onViewModeChange={mockOnViewModeChange}
        />
      </PaperProvider>
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it("renders correctly with server tab", () => {
    const { toJSON } = render(
      <PaperProvider theme={MD3LightTheme}>
        <TabPanels
          diagnostics={mockDiagnostics}
          extension={undefined}
          isLoading={false}
          selectedTab="server"
          viewMode="list"
          onLinkClick={mockOnLinkClick}
          onViewModeChange={mockOnViewModeChange}
        />
      </PaperProvider>
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
