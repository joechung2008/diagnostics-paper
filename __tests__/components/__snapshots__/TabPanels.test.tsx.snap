// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabPanels renders correctly in loading state 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  style={
    {
      "flex": 1,
    }
  }
>
  <View>
    ActivityIndicator: large
  </View>
</RCTScrollView>
`;

exports[`TabPanels renders correctly with build tab 1`] = `
<RCTScrollView
  style={
    {
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        {
          "padding": 10,
        }
      }
    >
      <View
        style={
          {
            "backgroundColor": "#FFFBFE",
            "padding": 10,
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "#FFFBFE",
              "borderColor": "#79747E",
              "borderWidth": 1,
            }
          }
        >
          <View
            style={
              {
                "backgroundColor": "#E7E0EC",
                "borderBottomColor": "#79747E",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "paddingBottom": 5,
                "paddingTop": 5,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#1C1B1F",
                  "flex": 1,
                  "fontFamily": "System",
                  "fontWeight": "bold",
                  "paddingLeft": 5,
                  "paddingRight": 5,
                }
              }
            >
              Name
            </Text>
            <Text
              style={
                {
                  "color": "#1C1B1F",
                  "flex": 1,
                  "fontFamily": "System",
                  "fontWeight": "bold",
                  "paddingLeft": 5,
                  "paddingRight": 5,
                }
              }
            >
              Value
            </Text>
          </View>
          <View
            style={
              {
                "borderBottomColor": "#79747E",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "paddingBottom": 5,
                "paddingTop": 5,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "minWidth": 0,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                  {
                    "color": "#1C1B1F",
                    "flex": 1,
                    "flexShrink": 1,
                    "fontFamily": "System",
                  },
                ]
              }
            >
              Build Version
            </Text>
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "minWidth": 0,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                  {
                    "color": "#1C1B1F",
                    "flex": 1,
                    "flexShrink": 1,
                    "fontFamily": "System",
                  },
                ]
              }
            >
              1.0.0
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`TabPanels renders correctly with error 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  style={
    {
      "flex": 1,
    }
  }
>
  <View>
    <Text
      style={
        [
          {
            "textAlign": "center",
          },
          {
            "color": "#BA1A1A",
          },
        ]
      }
    >
      Error loading diagnostics: 
      Test error
    </Text>
  </View>
</RCTScrollView>
`;

exports[`TabPanels renders correctly with extensions tab in list mode 1`] = `
<RCTScrollView
  style={
    {
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        {
          "padding": 10,
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "flex": 1,
            "opacity": 1,
            "transform": [
              {
                "scale": 1,
              },
              {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <RCTScrollView
          accessibilityLabel="Extensions"
          style={
            {
              "backgroundColor": "#FFFBFE",
              "flexGrow": 0,
              "flexShrink": 1,
              "minWidth": 0,
              "padding": 10,
            }
          }
        >
          <View>
            Button: Extension One
          </View>
        </RCTScrollView>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`TabPanels renders correctly with server tab 1`] = `
<RCTScrollView
  style={
    {
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        {
          "padding": 10,
        }
      }
    >
      <View
        style={
          {
            "backgroundColor": "#FFFBFE",
            "padding": 10,
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "#FFFBFE",
              "borderColor": "#79747E",
              "borderWidth": 1,
            }
          }
        >
          <View
            style={
              {
                "backgroundColor": "#E7E0EC",
                "borderBottomColor": "#79747E",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "paddingBottom": 5,
                "paddingTop": 5,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#1C1B1F",
                  "flex": 1,
                  "fontFamily": "System",
                  "fontWeight": "bold",
                  "paddingLeft": 5,
                  "paddingRight": 5,
                }
              }
            >
              Name
            </Text>
            <Text
              style={
                {
                  "color": "#1C1B1F",
                  "flex": 1,
                  "fontFamily": "System",
                  "fontWeight": "bold",
                  "paddingLeft": 5,
                  "paddingRight": 5,
                }
              }
            >
              Value
            </Text>
          </View>
          <View
            style={
              {
                "borderBottomColor": "#79747E",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "paddingBottom": 5,
                "paddingTop": 5,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "minWidth": 0,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                  {
                    "color": "#1C1B1F",
                    "flex": 1,
                    "flexShrink": 1,
                    "fontFamily": "System",
                  },
                ]
              }
            >
              Hostname
            </Text>
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "minWidth": 0,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                  {
                    "color": "#1C1B1F",
                    "flex": 1,
                    "flexShrink": 1,
                    "fontFamily": "System",
                  },
                ]
              }
            >
              test-host
            </Text>
          </View>
          <View
            style={
              {
                "borderBottomColor": "#79747E",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "paddingBottom": 5,
                "paddingTop": 5,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "minWidth": 0,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                  {
                    "color": "#1C1B1F",
                    "flex": 1,
                    "flexShrink": 1,
                    "fontFamily": "System",
                  },
                ]
              }
            >
              Uptime
            </Text>
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "minWidth": 0,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                  {
                    "color": "#1C1B1F",
                    "flex": 1,
                    "flexShrink": 1,
                    "fontFamily": "System",
                  },
                ]
              }
            />
          </View>
          <View
            style={
              {
                "borderBottomColor": "#79747E",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "paddingBottom": 5,
                "paddingTop": 5,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "minWidth": 0,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                  {
                    "color": "#1C1B1F",
                    "flex": 1,
                    "flexShrink": 1,
                    "fontFamily": "System",
                  },
                ]
              }
            >
              Server ID
            </Text>
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "minWidth": 0,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                  {
                    "color": "#1C1B1F",
                    "flex": 1,
                    "flexShrink": 1,
                    "fontFamily": "System",
                  },
                ]
              }
            >
              test-server
            </Text>
          </View>
          <View
            style={
              {
                "borderBottomColor": "#79747E",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "paddingBottom": 5,
                "paddingTop": 5,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "minWidth": 0,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                  {
                    "color": "#1C1B1F",
                    "flex": 1,
                    "flexShrink": 1,
                    "fontFamily": "System",
                  },
                ]
              }
            >
              Deployment ID
            </Text>
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "minWidth": 0,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                  {
                    "color": "#1C1B1F",
                    "flex": 1,
                    "flexShrink": 1,
                    "fontFamily": "System",
                  },
                ]
              }
            >
              test-deployment
            </Text>
          </View>
          <View
            style={
              {
                "borderBottomColor": "#79747E",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "paddingBottom": 5,
                "paddingTop": 5,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "minWidth": 0,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                  {
                    "color": "#1C1B1F",
                    "flex": 1,
                    "flexShrink": 1,
                    "fontFamily": "System",
                  },
                ]
              }
            >
              Node Versions
            </Text>
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "minWidth": 0,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                  {
                    "color": "#1C1B1F",
                    "flex": 1,
                    "flexShrink": 1,
                    "fontFamily": "System",
                  },
                ]
              }
            />
          </View>
          <View
            style={
              {
                "borderBottomColor": "#79747E",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "paddingBottom": 5,
                "paddingTop": 5,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "minWidth": 0,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                  {
                    "color": "#1C1B1F",
                    "flex": 1,
                    "flexShrink": 1,
                    "fontFamily": "System",
                  },
                ]
              }
            >
              Extension Sync | Total Sync All Count
            </Text>
            <Text
              style={
                [
                  {
                    "flex": 1,
                    "minWidth": 0,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                  },
                  {
                    "color": "#1C1B1F",
                    "flex": 1,
                    "flexShrink": 1,
                    "fontFamily": "System",
                  },
                ]
              }
            >
              5
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
